### PR TITLE
Improved release-tests automation using gauge

### PR DESCRIPTION
**Pull request by praveen4g0**
This Repo contains release-tests which Needs to run before every openshift-pipelines-operator release.

We use gauge framework to Automate release-test

know more about [Gauge](https://docs.gauge.org/index.html?os=linux&language=java&ide=vscode)

* Helps you to keep track of manual and Automated Tests in single Repo
* It performs tests on Real cluster 
* Validates features of tekton
   * pipelines
   * Cli (WIP)
   * Triggers (WIP)
   * operator
      * installing operators
      * upgrading operator (WIP)
      * deleting operator
   * catalog (WIP)

***Assumptions:***

*latest oc binary should be installed.

***includes***
- README.md file
- Makefile (helps you to download latest TKN binary)

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -4,39 +4,55 @@ Validation of OpenShift Pipeline releases
...
+  gauge install go
+  ```
+* Gauge html plugin
+  * can be installed using 
+  ```
+  gauge install html-report
+  ```  
+## Running tests 
+
+gauge run  --verbose --tag "e2e" specs/
```
Comment (ppitonak) [inline README.md]: `Error: unknown flag: --tag`

Comment (praveen4g0) [reply README.md]: typo it should be ```--tags``` updated

File: Makefile [hunk 2]
```diff
@@ -0,0 +1,23 @@
...
+	[[ -d "${DOWNLOAD_PATH}" ]] | mkdir -p ${DOWNLOAD_PATH}
+	curl -L -o ${DOWNLOAD_PATH}/tkn.tar.gz ${STABLE_DOWNLOAD_URL}
+	tar xvzf ${DOWNLOAD_PATH}/tkn.tar.gz -C ${DOWNLOAD_PATH} tkn
+	rm -rf ${DOWNLOAD_PATH}/tkn.tar.gz
+
+
+.PHONY: lint
+lint: ## run linter(s)
+	@echo "Linting..."
+	@golangci-lint run ./... --timeout 5m
```
Comment (ppitonak) [inline Makefile]: Did you forget to include a target for running the tests using containers?

**Comment by ppitonak** [comment]
Any idea why I cannot run it?
```
$ gauge run --env "test" --log-level=debug --verbose specs/olm
'gauge_environment' set to 'test'
'gauge_environment' set to 'default'
Plugin go is already installed.
Plugin html-report is already installed.
go: finding github.com/openshift/api v3.9.1-0.20190911180052-9f80b7806f58+incompatible
Checking updates...
Downloading https://downloads.gauge.org/plugin/go?l=go&p=go,html-report,screenshot,xml-report&o=linux&a=amd64
Downloading https://downloads.gauge.org/plugin/html-report?l=go&p=go,html-report,screenshot,xml-report&o=linux&a=amd64
Downloading https://downloads.gauge.org/plugin/screenshot?l=go&p=go,html-report,screenshot,xml-report&o=linux&a=amd64
Downloading https://downloads.gauge.org/plugin/xml-report?l=go&p=go,html-report,screenshot,xml-report&o=linux&a=amd64
can't load package: package main:
/tmp/gauge_temp1580736161599341842/gauge_test.go:5:4: invalid import path: "go:"
Failed to build project: Failed to compile project: exit status 1
Please ensure the project is in GOPATH.
Killing go runner.
Runner connection error: Runner with pid 35647 quit unexpectedly(exit status 1).
Error while killing runner: os: process already finished
Error ----------------------------------

[Gauge]
Failed to start gauge API: Runner with pid 35647 quit unexpectedly(exit status 1).

Get Support ----------------------------
        Docs:          https://docs.gauge.org
        Bugs:          https://github.com/getgauge/gauge/issues
        Chat:          https://gitter.im/getgauge/chat

Your Environment Information -----------
        linux, 1.0.6, 2bc49db
        go (0.1.3), html-report (4.0.8), screenshot (0.0.1), xml-report (0.2.2)
```

**Comment by praveen4g0** [comment]
> Any idea why I cannot run it?
> 
> ```
> $ gauge run --env "test" --log-level=debug --verbose specs/olm
> 'gauge_environment' set to 'test'
> 'gauge_environment' set to 'default'
> Plugin go is already installed.
> Plugin html-report is already installed.
> go: finding github.com/openshift/api v3.9.1-0.20190911180052-9f80b7806f58+incompatible
> Checking updates...
> Downloading https://downloads.gauge.org/plugin/go?l=go&p=go,html-report,screenshot,xml-report&o=linux&a=amd64
> Downloading https://downloads.gauge.org/plugin/html-report?l=go&p=go,html-report,screenshot,xml-report&o=linux&a=amd64
> Downloading https://downloads.gauge.org/plugin/screenshot?l=go&p=go,html-report,screenshot,xml-report&o=linux&a=amd64
> Downloading https://downloads.gauge.org/plugin/xml-report?l=go&p=go,html-report,screenshot,xml-report&o=linux&a=amd64
> can't load package: package main:
> /tmp/gauge_temp1580736161599341842/gauge_test.go:5:4: invalid import path: "go:"
> Failed to build project: Failed to compile project: exit status 1
> Please ensure the project is in GOPATH.
> Killing go runner.
> Runner connection error: Runner with pid 35647 quit unexpectedly(exit status 1).
> Error while killing runner: os: process already finished
> Error ----------------------------------
> 
> [Gauge]
> Failed to start gauge API: Runner with pid 35647 quit unexpectedly(exit status 1).
> 
> Get Support ----------------------------
>         Docs:          https://docs.gauge.org
>         Bugs:          https://github.com/getgauge/gauge/issues
>         Chat:          https://gitter.im/getgauge/chat
> 
> Your Environment Information -----------
>         linux, 1.0.6, 2bc49db
>         go (0.1.3), html-report (4.0.8), screenshot (0.0.1), xml-report (0.2.2)
> ```

Please do clone/setup this project under GOPATH, You should be able to run it

File: Makefile (hunk 2 above)
Comment (praveen4g0) [reply Makefile]: Yeah it's intentional, We are not supporting this feature as of now, but still you can run those steps which I have given manually using ```docker run -it```

**Comment by ppitonak** [comment]
@praveen4g0 I do have it on GOPATH, although it should not be necessary

File: Dockerfile [hunk 3]
```diff
@@ -0,0 +1,57 @@
...
+		| tar -v -C /usr/local -xz
+
+ENV PATH /usr/local/go/bin:$PATH
+RUN mkdir -p /go/src /go/bin && chmod -R 777 /go
+ENV GOROOT /usr/local/go
+ENV GOPATH /go
+ENV PATH /go/bin:$PATH
+
+# install oc
+RUN wget https://github.com/openshift/origin/releases/download/v3.11.0/openshift-origin-client-tools-v3.11.0-0cbc58b-linux-64bit.tar.gz && \
```
Comment (ppitonak) [inline Dockerfile]: Why 3.11? Is Tekton/Pipelines supported on 3.11?

**Comment by ppitonak** [comment]
@praveen4g0 I see the following inside container:

```
Runner connection error: Timed out connecting to 127.0.0.1:46869
Error ----------------------------------

[Gauge]
Failed to start gauge API: Timed out connecting to 127.0.0.1:46869

Get Support ----------------------------
        Docs:          https://docs.gauge.org
        Bugs:          https://github.com/getgauge/gauge/issues
        Chat:          https://gitter.im/getgauge/chat

Your Environment Information -----------
        linux, 1.0.6, 2bc49db
        go (0.1.3), html-report (4.0.8), screenshot (0.0.1)
```

File: Dockerfile (hunk 3 above)
Comment (praveen4g0) [reply Dockerfile]: @sthaha we better remove Dockerfile as we don't support it, what do you say?

@ppitonak Good catch that was just work around as I said, basically we are using oc client  in some scenarios so technically it should work. But good catch Need to update that to latest version.

**Comment by praveen4g0** [comment]
> @praveen4g0 I see the following inside container:
> 
> ```
> Runner connection error: Timed out connecting to 127.0.0.1:46869
> Error ----------------------------------
> 
> [Gauge]
> Failed to start gauge API: Timed out connecting to 127.0.0.1:46869
> 
> Get Support ----------------------------
>         Docs:          https://docs.gauge.org
>         Bugs:          https://github.com/getgauge/gauge/issues
>         Chat:          https://gitter.im/getgauge/chat
> 
> Your Environment Information -----------
>         linux, 1.0.6, 2bc49db
>         go (0.1.3), html-report (4.0.8), screenshot (0.0.1)
> ```
How do I reproduce it ?? Basically , you need to perform oc login inside container as it tries to connect to running openshift cluster

**Comment by ppitonak** [comment]
> > @praveen4g0 I see the following inside container:
> > ```
> > Runner connection error: Timed out connecting to 127.0.0.1:46869
> > Error ----------------------------------
> > 
> > [Gauge]
> > Failed to start gauge API: Timed out connecting to 127.0.0.1:46869
> > 
> > Get Support ----------------------------
> >         Docs:          https://docs.gauge.org
> >         Bugs:          https://github.com/getgauge/gauge/issues
> >         Chat:          https://gitter.im/getgauge/chat
> > 
> > Your Environment Information -----------
> >         linux, 1.0.6, 2bc49db
> >         go (0.1.3), html-report (4.0.8), screenshot (0.0.1)
> > ```
> 
> How do I reproduce it ?? Basically , you need to perform oc login inside container as it tries to connect to running openshift cluster

```
podman build -t ppitonak/release-tests .
podman run --rm -it --net=host -P --env TKN_VERSION=0.6.0 --env USER_NAME=kubeadmin --env PASSWORD=*** --env CLUSTER_URL=https://api.***:6443 ppitonak/release-tests
gauge run --env "test" --log-level=debug --verbose specs/olm
```

**Comment by ppitonak** [comment]
The problem was that I didn't have `tkn` in build directory. Please fix the tests to explicitly say what's wrong instead of misleading timeout error.

File: Dockerfile (hunk 3 above)
Comment (ppitonak) [reply Dockerfile]: @praveen4g0 what do you mean by "don't support" it. I suggest to promote using containers for local execution and development of tests.

Comment (sthaha) [reply Dockerfile]: by don't support he means that it is a problem can solve later. The value currently is in writing tests and be able to run tests which you now can locally without having to rely on docker. Being able to use a container that has all tools packages IMO is only a good to have at this point.

**Pull request merged**